### ISSUE TITLE
Fonts not loading in page/site editor with Gutenberg 15.9 RC1

### ISSUE THREAD

**Issue by bgardner** [WordPress/gutenberg#50965]
### Description

I have several themes that load Outfit font from Google, and with Gutenberg 15.9 RC1 active, they are not being enqueued properly in page/site editors. The front end works as expected.  To further test, I confirmed this is also true of the Hey theme by Automattic. If I deactivate the plugin, the fonts load as expected.

### Step-by-step reproduction instructions

1. Activate theme with Google fonts
2. Activate Gutenberg 15.9 RC1
3. View content on page/site editors

### Environment info

WordPress 6.2.2, Gutenberg 15.9 RC1

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by bgardner** [comment]
Adding video to show what happens when plugin is activated/deactivated:

https://github.com/WordPress/gutenberg/assets/486261/7b485f14-f3c8-4447-a8c2-38384c791ea9

**Comment by ndiego** [comment]
I am able to confirm with TT3 using the included font families. cc @c4rl0sbr4v0 this would be great to get fixed before 15.9 goes out the door.

**Comment by WraithKenny** [comment]
Able to reproduce also, shows Helvetica in FSE editor.

**Comment by WraithKenny** [comment]
#50929 is also a font related regression.

**Comment by cbravobernal** [comment]
I'm not really sure that I will be able to fix it before the RC2 😢

**Comment by richtabor** [comment]
Confirmed this issue.

**Comment by hellofromtonya** [comment]
>Activate theme with Google fonts

Does the theme have a local asset file for the Google font? Or is a custom provider included that remotely interacts with Google Fonts foundry API?

Why am I asking?

Because Gutenberg ships with a "local" provider for processing local asset files. But if the `@font-face` src comes from Google Fonts API, then a custom provider is needed, which could mean there's an issue within that provider (which is not provided in Gutenberg).

Knowing this can help with reproduction and investigation.

**Comment by hellofromtonya** [comment]
# Test Report

Env:
* Plugins: Gutenberg 15.9 RC2
* Theme: TT3
* Must-Use: none
* LocalHost: Local
* OS: macOS
* Browser: Firefox

## Test Steps

1. Downloaded the [15.9 RC2 gutenberg.zip file](https://github.com/WordPress/gutenberg/releases/tag/v15.9.0-rc.2) and added it as new plugin through WP's Plugins > Add New UI
2. Navigated to the Site Editor > Styles > Typography UI
3. Selected "Headings"
4. Selected the "Font" as "Source Serif Pro" (which has a local font file in the TT3 theme)
5. Compared the rendered heading font to the [same font on Google Fonts site](https://fonts.google.com/specimen/Source+Serif+Pro?preview.text=Mindblown:%20a%20blog%20about%20philosophy.&preview.text_type=custom)

## Results

<img width="2416" alt="Screen Shot 2023-05-31 at 10 29 36 AM" src="https://github.com/WordPress/gutenberg/assets/7284611/3b72e38c-cbf4-4c4e-8d60-1fd3c0cbbb58">

✅ Font rendered as expected. 

❌ Could not reproduce the reported issue using a font that has a local asset file in the theme.

## Summary

Next, I'll try using a Google Fonts Provider that interacts with Google Fonts Foundry API rather than supplying a local font asset file. I'll use [this test plugin](https://github.com/ironprogrammer/webfonts-jetpack-test) that includes [Jetpack's Google Fonts Provider package](https://github.com/Automattic/jetpack-google-fonts-provider/tree/trunk).

**Comment by hellofromtonya** [comment]
# Test Report 2

Env:
* Plugins:
    * Gutenberg 15.9 RC2
    * [Webfonts Jetpack Test](https://github.com/ironprogrammer/webfonts-jetpack-test) which uses [Jetpack Google Fonts Provider package](https://packagist.org/packages/automattic/jetpack-google-fonts-provider)
* Theme: TT3
* Must-Use: none
* LocalHost: Local
* OS: macOS
* Browser: Firefox

## Test Steps

### Set up

1. Gutenberg 15.9 RC plugin:
     * Downloaded the [15.9 RC2 gutenberg.zip file](https://github.com/WordPress/gutenberg/releases/tag/v15.9.0-rc.2)
     * Added it as new plugin through WP's Plugins > Add New UI
     * Activated it.
2. [Webfonts Jetpack Test plugin](https://github.com/ironprogrammer/webfonts-jetpack-test)
   *  Downloaded the zip file from GitHub by clicking on the "Code" selector and then selecting "Download ZIP".
   * Added it as new plugin through WP's Plugins > Add New UI.
   * Activated it.

### Scenario 1: Selecting a Google Font in Site Editor but not saving the custom style.

1. Navigated to the Site Editor > Styles > Typography UI
2. Selected "Headings"
3. Selected the "Font" as "Merriweather" (which is a Google Font that does not have a local asset file in the plugin or theme but instead will interact with Google Fonts Foundry API)
4. Selected the "Appearance" as "Bold Italic".
5. Compared the rendered heading font to the [same font on Google Fonts site](https://fonts.google.com/specimen/Merriweather?preview.text=Mindblown:%20a%20blog%20about%20philosophy.&preview.text_type=custom). Do they match?

### Scenario 2: After saving the custom style in the Site Editor.

1. Save the custom heading.
2. Reload the Site Editor.
3. Again compare the headings to the [same font on Google Fonts site](https://fonts.google.com/specimen/Merriweather?preview.text=Mindblown:%20a%20blog%20about%20philosophy.&preview.text_type=custom)
Do they match?

### Scenario 3: Editing an existing post after saving the custom style in the Site Editor.

1. Exit the Site Editor.
2. Open an existing Post by navigating to All Posts > selecting the Hello World or other existing post
3. Again compare the headings to [Google Fonts site](https://fonts.google.com/specimen/Merriweather?preview.text=Hello%20World&preview.text_type=custom). Do they match?

## Results

### Scenario 1:

<img width="2421" alt="Screen Shot 2023-05-31 at 10 32 45 AM" src="https://github.com/WordPress/gutenberg/assets/7284611/bdf9177a-904a-43fb-8d16-1f53bb0865fc">


✅ Google Font rendered as expected. 

❌ Could not reproduce the reported issue.

### Scenario 2: After saving the custom style in the Site Editor.

<img width="2049" alt="Screen Shot 2023-05-31 at 10 41 12 AM" src="https://github.com/WordPress/gutenberg/assets/7284611/4f3fa117-842a-4731-bf4d-de68c419216c">

✅ Google Font rendered as expected. 

❌ Could not reproduce the reported issue.

### Scenario 3: Editing an existing post after saving the custom style in the Site Editor.

<img width="1648" alt="Screen Shot 2023-05-31 at 10 43 30 AM" src="https://github.com/WordPress/gutenberg/assets/7284611/14ba6b2b-c016-4afa-9640-7ad4706013be">

✅ Google Font rendered as expected. 

❌ Could not reproduce the reported issue.

## Summary

Hmm, in all 3 scenarios, I could not reproduce the reported issue for a Google Font. Each scenario rendered as expected and the `@font-face` styles were correct in the HTML for each.

**Comment by bgardner** [comment]
@hellofromtonya Yes, my themes (as did the Hey theme I also tested) come packaged with local font source files. For example, you can see [how Frost has them included](https://github.com/wpengine/frost/tree/trunk/assets/fonts).

**Comment by bgardner** [comment]
@hellofromtonya And additionally, [how they are registered via theme.json](https://github.com/wpengine/frost/blob/trunk/theme.json#L184).

**Comment by hellofromtonya** [comment]
Thanks @bgardner. Let me try the Frost theme with my test set up. Test Report coming.

**Comment by hellofromtonya** [comment]
# Test Report

Env:
* Plugins: Gutenberg 15.9 RC2
* Theme: Frost
* Must-Use: none
* LocalHost: Local
* OS: macOS
* Browser: Firefox

## Test Steps

1. Gutenberg 15.9.0: install and activate.
2. Theme: Appearance > Themes > Add New > find Frost. Then install and activate.
3. Navigate to Appearance > Editor
4. Compare the heading "Experience the next generation of WordPress" to the [Outfit font on the Google Fonts site](https://fonts.google.com/specimen/Outfit?preview.text=Experience%20the%20next%20generation%20of%20WordPress.&preview.text_type=custom). Do they match?
5. Using your browser's dev tools, inspect the HTML. Search for `wp-fonts-local`. Does the CSS match:

```html
<style id="wp-fonts-local">
@font-face{font-family:Outfit;font-style:normal;font-weight:100 900;font-display:block;src:url('/wp-content/themes/frost/assets/fonts/Outfit-Variable.woff2') format('woff2');font-stretch:normal;}
</style>
```

## Results

### Without Gutenberg activated


✅  The `@font-face` styles are correct

```html
<style id="wp-block-library-inline-css">
@font-face{font-family:Outfit;font-style:normal;font-weight:100 900;font-display:block;src:url('/wp-content/themes/frost/assets/fonts/Outfit-Variable.woff2') format('woff2');font-stretch:normal;}
</style>
```

✅  The font matches the Outfit rendering on the Google Fonts site
<img width="2220" alt="Screen Shot 2023-05-31 at 1 20 47 PM" src="https://github.com/WordPress/gutenberg/assets/7284611/5bad892c-ee6a-41fa-94da-aee16e105eb3">

### With Gutenberg 15.9.0 activated

✅  The `@font-face` styles are correct

```html
<style id="wp-fonts-local">
@font-face{font-family:Outfit;font-style:normal;font-weight:100 900;font-display:block;src:url('/wp-content/themes/frost/assets/fonts/Outfit-Variable.woff2') format('woff2');font-stretch:normal;}
</style>
```

❌ The font does not match the Outfit rendering on the Google Fonts site

<img width="2262" alt="Screen Shot 2023-05-31 at 1 16 58 PM" src="https://github.com/WordPress/gutenberg/assets/7284611/134e1285-7940-457c-a887-e084fb6b86f6">

## Summary

Yes, I can reproduce the issue using the Frost theme.

👉 But the `@font-face` CSS generated is the same with and without Gutenberg plugin activated.

Hmm, why isn't the Outfit font being rendered when it's `@font-face` CSS is there?

**Comment by hellofromtonya** [comment]
Investigating ...

## Is the Fonts API working?

Yes.

As I noted in my last test report, the `@font-face` CSS generated is the same with and without Gutenberg plugin activated:

With the plugin activated:
```html
<style id="wp-fonts-local">
@font-face{font-family:Outfit;font-style:normal;font-weight:100 900;font-display:block;src:url('/wp-content/themes/frost/assets/fonts/Outfit-Variable.woff2') format('woff2');font-stretch:normal;}
</style>
```

Without the plugin activated:
```html
<style id="wp-block-library-inline-css">
@font-face{font-family:Outfit;font-style:normal;font-weight:100 900;font-display:block;src:url('/wp-content/themes/frost/assets/fonts/Outfit-Variable.woff2') format('woff2');font-stretch:normal;}
</style>
```

The Fonts API is working ✅ 

## What about which font-family is assigned to the `h2` element?

Inspecting the element and its rules:

With the Gutenberg plugin activated:
```css
.editor-styles-wrapper {
  --wp--style--root--padding-top: 0px;
  --wp--style--root--padding-right: var(--wp--custom--spacing--gap);
  --wp--style--root--padding-bottom: 0px;
  --wp--style--root--padding-left: var(--wp--custom--spacing--gap);
  color: var(--wp--preset--color--contrast);
  font-family: var(--wp--preset--font-family--primary);
}
```

where  `--wp--preset--font-family--primary = Outfit, sans-serif`.

<img width="2021" alt="Screen Shot 2023-05-31 at 1 51 58 PM" src="https://github.com/WordPress/gutenberg/assets/7284611/4e389062-95dd-4a74-bb80-627fd1cb5d28">

* Without the Gutenberg plugin activated:

It's the same.

![Screen Shot 2023-05-31 at 1 53 07 PM](https://github.com/WordPress/gutenberg/assets/7284611/2f0027ff-0150-4bde-8fea-2b1f2efeb01c)

```css
.editor-styles-wrapper {
  --wp--style--root--padding-top: 0px;
  --wp--style--root--padding-right: var(--wp--custom--spacing--gap);
  --wp--style--root--padding-bottom: 0px;
  --wp--style--root--padding-left: var(--wp--custom--spacing--gap);
  color: var(--wp--preset--color--contrast);
  font-family: var(--wp--preset--font-family--primary);
}
```

where  `--wp--preset--font-family--primary = Outfit, sans-serif`.

## What next?

Picture me shrugging 🤷‍♀️  Hmm, likely someone needs to confirm my findings. Do you also get the same `@font-face` CSS with and without the Gutenberg plugin activated?

**Comment by bgardner** [comment]
@hellofromtonya Worth noting, I know you tested the H2 in this, but it's all type—paragraphs, etc.

**Comment by ndiego** [comment]
I also wanted to mention you can test this with fonts included in TT3. In Global Styles I set all text to IBM Plex Mono. This is included with the theme and I do not have it on my local machine. With 15.9, this font is no longer loaded. 

**How it should look (Gutenberg deactivated):** 
<img width="1683" alt="image" src="https://github.com/WordPress/gutenberg/assets/4832319/5e145727-e5a6-4302-b599-e3b0a204f522">

<img width="508" alt="image" src="https://github.com/WordPress/gutenberg/assets/4832319/3a53eda8-7c7e-4bc5-b9c0-769232c0c56e">


**How it looks with Gutenberg 15.9:** 
<img width="1683" alt="image" src="https://github.com/WordPress/gutenberg/assets/4832319/f953a71d-4d57-4e61-a781-3f03f4824c7d">

<img width="493" alt="image" src="https://github.com/WordPress/gutenberg/assets/4832319/c2d82a84-a9e6-4d3b-a066-0a01f4837111">

**Comment by hellofromtonya** [comment]
Thank you @bgardner! I see the issue. Let me explain.

## tl;dr
In the admin area, the font `src` needs to be absolute URL to work within an iframe.

## Explanation:

The issue:
The `WP_Fonts_Provider_Local::compile_src()` converts the font src `'url'` from absolute to relative. Relative works fine when the font asset files are referenced in the main document. But in an iframe, the relative URL relates to the main document, rather than the path from the iframe to the font asset file.

When making the URL absolute, notice the asset file is included in the `editor-canvas` (the iframe) list.
<img width="609" alt="editor-canvas source files" src="https://github.com/WordPress/gutenberg/assets/7284611/64222169-d0d5-471e-8b2f-790e2039212c">

## Possible solution:
There are a couple of ways to solve this:

* Option 1: Within an iframed editor, use the absolute URL.
* Option 2: Remove the make relative code so that all font src URLs are absolute.

I don't remember the reasoning for why relative URLs were used in the Local Provider. Pinging @aristath for insights.

But I think Option 2 is likely the best option. Why? It will ensure assets load in all instances (iframe or main document) while reducing the amount of code. That gives a bonus of less code to maintain and a tiny micro-optimization for performance (less code to run).

I'll open a PR for Option 2 for consideration. I would like @aristath input to refresh my memory of why relative URLs were used, i.e. in case they are indeed needed in certain instances.

**Comment by ndiego** [comment]
Thanks for the insight @hellofromtonya . @ellatrix do you think this issue in 15.9 could be related to https://github.com/WordPress/gutenberg/pull/50875? This only started occurring in 15.9.

**Comment by hellofromtonya** [comment]
@bgardner Would you have time to test the bugfix? If no, that's okay. But if yes, please test PR https://github.com/WordPress/gutenberg/pull/51178 with the Frost theme. Testing instructions are listed in the description. It works for me. I want to certain it also works for you.

**Comment by WraithKenny** [comment]
I think this would also be testable on twenty-twentythree, since it also loads local fonts.

**Comment by hellofromtonya** [comment]
>I think this would also be testable on twenty-twentythree, since it also loads local fonts.

You're right @WraithKenny, though there's more steps for testing as you need to select one of the fonts in Site Editor > Styles > Typography UI. For example you can select "Source Serif Pro" for all the Text fonts.

**Comment by webd-uk** [comment]
OMG, I've been stuck on exactly this issue for days. THANK YOU!
